pkg/git/provider_bitbucket-server.go

```diff
@@ -46,22 +54,82 @@ func NewBitbucketServerProvider(baseURL string) (Provider, error) {
 	}, nil
 }
 
-func (bbs *bitbucketServer) Type() ProviderType {
-	return bbs.providerType
-}
-
 func (bbs *bitbucketServer) BaseURL() string {
 	urlClone := *bbs.apiURL
 	urlClone.Path = ""
 	urlClone.RawQuery = ""
 	return urlClone.String()
 }
 
-func (bbs *bitbucketServer) VerifyToken(ctx context.Context, tokenType TokenType, token string) error {
-	log.G(ctx).Debug("Skip verifying token for bitbucket, to be implemented later")
+func (bbs *bitbucketServer) SupportsMarketplace() bool {
+	return false
+}
+
+func (bbs *bitbucketServer) Type() ProviderType {
+	return bbs.providerType
+}
+
+func (bbs *bitbucketServer) VerifyRuntimeToken(ctx context.Context, token string) error {
+	return bbs.checkProjectAdminPermission(ctx, token)
+}
+
+func (bbs *bitbucketServer) VerifyUserToken(ctx context.Context, token string) error {
+	return bbs.checkRepoReadPermission(ctx, token)
+}
+
+func (bbs *bitbucketServer) checkProjectAdminPermission(ctx context.Context, token string) error {
+	username, err := bbs.getCurrentUsername(ctx, token)
+	if err != nil {
+		return err
+	}
+
+	urlPath := fmt.Sprintf("users/%s/repos", username)
+	body := &createRepoBody{
+		Name: "!invalid",
+	}
+	res, err := bbs.request(ctx, token, http.MethodPost, urlPath, body)
```

Comment:
POST to `users/<username>/repos` with an invalid repo name.
if it returns 400 - the token has "Project admin" permission
if it returns 403 - the token does not have the permission

Reply:
Please write this github comment inside the code